### ISSUE TITLE
Configurable registry for binary downloads

### ISSUE THREAD

**Issue by chadhietala** [evanw/esbuild#1193]
We use an internal artifactory instance that contains all of the packages we we need to build and those machines don't have access to the internet. Currently when we try to use esbuild it fails in CI since the installer attempts to download the public registry.

Would you be open to a PR that added npmrc support for defining the registry? I was thinking something like the [`sass_binary_site`](https://www.npmjs.com/package/node-sass#binary-configuration-parameters) config.

**Comment by evanw** [comment]
There is already a npm config for this: https://docs.npmjs.com/configuring-your-registry-settings-as-an-npm-enterprise-user. I remember testing this in the past using https://verdaccio.org/ and it seemed to work. Setting `registry` in your `.npmrc` should set the `npm_config_registry` environment variable which should persist into esbuild's nested npm invocation. Have you tried doing that already? If so, why didn't it work?

**Comment by chadhietala** [comment]
TL;DR this is a Yarn problem on how it applies the npmrc. There is a work around (see below) but ideally I need to get off of Yarn.

Having dived in to this a bit more there are a couple things I think going on. We rely on the client; either npm or yarn, to read the registry the npmrc and we do not proxy all calls to public repo back internally into the private instance. This is largely a security policy on our end. 

That said I think this is related to the fact that we pre-Yarn 2.0 as the initial installer and it has issues with apply `.npmrc` onto the `process.env`. I confirmed with verdaccio that doing:

```bash
yarn install # hits public npm registry without going through proxy
```

```bash
npm_config_registry="http://localhost:5555" yarn install # hits public npm registry without going through proxy
```

```bash
NPM_CONFIG_REGISTRY="http://localhost:5555" yarn install # goes through proxy
```

```bash
npm install #  goes through proxy
```

**Comment by SkReD** [comment]
Hi @evanw! Thanks for the great tool! Faced this problem from the topic but with yarn 1 and none of listed solutions worked:
1. `export NPM_CONFIG_REGISTRY`
2. `export npm_config_registry`
3. `yarn config set registry`
4. `yarn config set registry -g`

Yarn issues https://github.com/yarnpkg/yarn/issues/5854 and https://github.com/yarnpkg/yarn/issues/5683

Only worked solution was pre-install packages via `npm install` with `--registry` option for cache warmup. Also in docker there seems to be problem with `process.env.TMPDIR` and that `require('os').tmpdir()` is returns `cwd` instead of temp directory path (https://github.com/npm/npm/issues/4531).

Can install script be extended with support for custom env variable for registry and with `--unsafe-perm` flag for correct `process.env.TMPDIR` value?
